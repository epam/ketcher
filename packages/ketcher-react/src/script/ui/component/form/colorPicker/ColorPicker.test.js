import React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'

import ColorPicker from './ColorPicker'

const openPreset = async () => {
  const presetToggleBtn = screen.getByTestId('color-picker-preview')
  userEvent.click(presetToggleBtn)
  const preset = await screen.findByTestId('color-picker-preset')
  await waitFor(() => expect(preset).toBeInTheDocument())
}
const openPalette = async () => {
  const pickerToggleBtn = screen.getByTestId('color-picker-btn')
  userEvent.click(pickerToggleBtn)
  const palette = await screen.findByTestId('color-palette')
  await waitFor(() => expect(palette).toBeInTheDocument())
}

describe('should be rendered correctly', () => {
  it('should be rendered without crashing', () => {
    const { asFragment } = render(<ColorPicker value="#000000" />)
    expect(asFragment(<ColorPicker />)).toMatchSnapshot()
  })
})

describe('should toggle color picker dialog', () => {
  it('should show color preset on click', async () => {
    const { container } = render(<ColorPicker />)
    await openPreset()
    expect(
      container.getElementsByClassName('classes.colorPickerWrap')
    ).toBeDefined()
  })

  it('should show color picker dialog on click', async () => {
    const { container } = render(<ColorPicker />)
    await openPreset()
    await openPalette()
    expect(container.getElementsByClassName('react-colorful')[0]).toBeDefined()
  })

  it('should hide color picker dialog on click outside picker', async () => {
    const { container } = render(<ColorPicker />)
    await openPreset()
    await openPalette()
    const overlay = screen.getByTestId('color-picker-field-open')
    await new Promise((resolve) => {
      setTimeout(() => {
        userEvent.click(overlay)
        resolve()
      }, 300)
    })
    expect(await screen.findByTestId('color-picker-field')).toBeInTheDocument()

    expect(
      container.getElementsByClassName('react-colorful')[0]
    ).toBeUndefined()
  })
})

describe('should pick color correctly', () => {
  it('should call onChange callback with picked color', async () => {
    const onChange = jest.fn()
    render(<ColorPicker onChange={onChange} />)
    await openPreset()
    await openPalette()
    const colorInput = screen.getByTestId('color-picker-input')
    fireEvent.change(colorInput, { target: { value: '#4d4d4d' } })
    expect(onChange).toBeCalledWith('#4d4d4d')
  })

  it('should display picked color correctly', () => {
    render(<ColorPicker value="#000000" />)
    expect(
      screen.getByTestId('color-picker-preview').style.backgroundColor
    ).toBe('rgb(0, 0, 0)')
  })
})
