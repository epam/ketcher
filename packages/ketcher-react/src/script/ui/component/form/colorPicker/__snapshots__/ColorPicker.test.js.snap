// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be rendered correctly should be rendered without crashing 1`] = `
<DocumentFragment>
  <div
    class="colorPickerInput"
  >
    <div
      class="colorPickerBtn"
      data-testid="color-picker-btn"
    >
      #000000
      <span
        class="colorPickerPreview"
        data-testid="color-picker-preview"
        style="background-color: rgb(0, 0, 0);"
      />
    </div>
    <div
      class="colorPickerWrap"
    />
  </div>
</DocumentFragment>
`;
